src/lib/Live/Eval.hs

```diff
@@ -6,46 +6,87 @@
 
 module Live.Eval (RFragment (..), SetVal(..), watchAndEvalFile) where
 
-import Control.Concurrent (forkIO, readChan, threadDelay)
+import Control.Concurrent (forkIO, killThread, readChan, threadDelay, ThreadId)
 import Control.Monad.Reader
 import Control.Monad.State.Strict
 import qualified Data.Map.Strict as M
 
 import Data.Aeson (ToJSON, toJSON, (.=))
 import qualified Data.Aeson as A
+import Data.Text.Prettyprint.Doc
 import System.Directory (getModificationTime)
 
 import Actor
 import Parser
 import RenderHtml (ToMarkup, pprintHtml)
 import Syntax
 import TopLevel
+import Util (onFst, onSnd)
 
 type NodeId = Int
 data WithId a = WithId { getNodeId :: NodeId
                        , withoutId :: a }
+                deriving Show
+
 data RFragment = RFragment (SetVal [NodeId])
                            (M.Map NodeId SourceBlock)
                            (M.Map NodeId Result)
 
 -- Start watching and evaluating the given file.  Returns a channel on
 -- which one can subscribe to updates to the evaluation state.
+--
+-- The overall system looks like this:
+-- - `forkWatchFile` creates an actor that watches the file for
+--   changes and sends `FileChanged` messages to the driver.
+-- - `runDriver` creates the main driver actor, which manages
+--   the evaluation state and produces rendering fragments.
+-- - `logServer` creates an actor that accumulates rendering fragments
+--   from the driver and broadcasts them to any subscribed clients.
+--
+-- `FileChanged` messages from the watch file actor may invalidate the
+-- current state.  The driver delegates the actual evaluation to a
+-- sub-thread so it can remain responsive.
+--
+-- `watchAndEvalFile` returns the channel by which a client may
+-- subscribe by sending a write-only view of its input channel.
 watchAndEvalFile :: FilePath -> EvalConfig -> TopStateEx
                  -> IO (PChan (PChan RFragment))
 watchAndEvalFile fname opts env = do
   (_, resultsChan) <- spawn logServer
   let cfg = (opts, subChan Publish resultsChan)
-  (_, sourceChan) <- spawn $ runDriver cfg env
-  forkWatchFile fname sourceChan
+  (_, driverChan) <- spawn $ runDriver cfg env
+  forkWatchFile fname $ subChan FileChanged driverChan
```

Comment:
This `subChan` business is really awkward. We should really make channels functors, because that's what it represents I think? No need to do it in this PR, but just a thought I had reading this

Reply:
They're not `Functor`s because they're contravariant: `subChan :: (a -> b) -> (PChan b) -> (PChan a)`.  But there is probably a class for this pattern.